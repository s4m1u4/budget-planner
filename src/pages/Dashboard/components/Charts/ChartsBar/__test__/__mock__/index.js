export { TITLE, DATA } from "./mock";
