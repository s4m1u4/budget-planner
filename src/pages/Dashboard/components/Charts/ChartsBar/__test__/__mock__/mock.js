export const TITLE = "Income";
export const DATA = [
  { name: "Work", income: 17369, expense: -13223 },
  { name: "Travel", income: 74123, expense: -84534 },
  { name: "Education", income: 49763, expense: -2533 },
];
