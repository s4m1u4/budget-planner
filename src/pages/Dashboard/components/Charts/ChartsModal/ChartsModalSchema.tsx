import * as yup from "yup";

export const ChartsModalSchema = yup.object({
  id: yup.string().required("Category is required"),
});
