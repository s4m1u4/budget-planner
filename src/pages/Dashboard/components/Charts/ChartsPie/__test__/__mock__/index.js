export { TITLE, DATA_INCOME } from "./mock";
