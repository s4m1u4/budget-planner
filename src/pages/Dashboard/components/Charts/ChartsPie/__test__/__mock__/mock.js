export const TITLE = "Income";
export const DATA_INCOME = [
  { name: "Work", value: 17369 },
  { name: "Travel", value: 106787 },
  { name: "Education", value: 10456 },
];
