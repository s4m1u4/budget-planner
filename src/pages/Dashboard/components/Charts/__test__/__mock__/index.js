export { CATEGORIES, RECORDS } from "./mock";
