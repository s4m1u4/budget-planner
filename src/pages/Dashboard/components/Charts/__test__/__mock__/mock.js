export const RECORDS = [
  {
    id: "9cc6c0d7-24e3-46aa-aa7c-73e86c1dd230",
    amount: "2354.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
  {
    id: "bc732adb-7b27-4d76-944c-4e7013eb97fa",
    amount: "836.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
  {
    id: "9410db09-8e1c-4c76-a6b3-e766d9b53ae5",
    amount: "2454.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
];

export const CATEGORIES = [
  {
    id: "084a3558-6b14-4e38-9d24-47e912021384",
    title: "Work",
    description: "Work",
    isDefault: false,
  },
  {
    id: "0ce314b9-a547-4f57-9d64-e466777ba8e9",
    title: "Travel",
    description: "awdawda",
    isDefault: false,
  },
  {
    id: "ea298d16-8bdb-4c5b-8198-6b2e5ce2f9a2",
    title: "Education",
    description: "Education it's so important for you!",
    isDefault: false,
  },
];
