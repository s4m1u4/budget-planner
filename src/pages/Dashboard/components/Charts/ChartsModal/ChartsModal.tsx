import React, { FC } from "react";
import { useFormik } from "formik";
import { ChartsModalSchema } from "./ChartsModalSchema";
import { ModalComponent } from "components/shared";
import { Button, MenuItem, TextField } from "@mui/material";
import { ICategory } from "types";

import { ButtonGroup, Title } from "./ChartsModal.styles";

interface ChartsModalProps {
  open: boolean;
  onSubmit: (values: IValuesData) => void;
  handleClose: () => void;
  categories: ICategory[];
  deleteCategory: (id: string) => void;
}

interface IValuesData {
  id: string;
}

export const ChartsModal: FC<ChartsModalProps> = ({
  open,
  onSubmit,
  handleClose,
  categories,
  deleteCategory,
}) => {
  const handleSubmit = async (values: IValuesData) => {
    formik.resetForm();
    await deleteCategory(values.id);
    await onSubmit(values);
    handleClose();
  };

  const formik = useFormik({
    initialValues: {
      id: "",
    },
    validationSchema: ChartsModalSchema,
    onSubmit: handleSubmit,
  });

  return (
    <ModalComponent open={open} onClose={handleClose}>
      <Title>Delete category</Title>
      <form onSubmit={formik.handleSubmit}>
        <TextField
          select
          fullWidth
          sx={{ marginBottom: "15px" }}
          name="id"
          size="small"
          label="Category"
          value={formik.values.id}
          onBlur={formik.handleBlur}
          onChange={formik.handleChange}
          error={formik.touched.id && !!formik.errors.id}
          helperText={formik.touched.id && formik.errors.id}
        >
          {categories.map((category) => (
            <MenuItem key={category.id} value={category.id}>
              {category.title}
            </MenuItem>
          ))}
        </TextField>
        <ButtonGroup>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="error"
            type="submit"
          >
            Delete category
          </Button>
          <Button
            sx={{ color: "#fff" }}
            variant="contained"
            color="primary"
            onClick={handleClose}
          >
            Close
          </Button>
        </ButtonGroup>
      </form>
    </ModalComponent>
  );
};
