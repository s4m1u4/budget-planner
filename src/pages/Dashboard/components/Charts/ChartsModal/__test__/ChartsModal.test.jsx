import { ChartsModal } from "../ChartsModal";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CATEGORIES } from "./__mock__";

const renderComponent = () => {
  return render(
    <ChartsModal
      open={open}
      onSubmit={onSubmit}
      categories={CATEGORIES}
      handleClose={handleClose}
      deleteCategory={deleteCategory}
    />
  );
};

const renderComponentAsync = async () => {
  await act(
    async () =>
      await render(
        <ChartsModal
          open={open}
          onSubmit={onSubmit}
          categories={CATEGORIES}
          handleClose={handleClose}
          deleteCategory={deleteCategory}
        />
      )
  );
};

const open = true;
const onSubmit = jest.fn();
const handleClose = jest.fn();
const deleteCategory = jest.fn();

describe("ChartsModal component", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderComponent();
    });

    it("the title", () => {
      expect(
        screen.getByRole("heading", { name: /delete category/i })
      ).toBeInTheDocument();
    });

    it('an input with name="category"', () => {
      expect(
        screen.getByRole("button", { name: /category ​/i })
      ).toBeInTheDocument();
    });

    it('a button "Delete category"', () => {
      expect(
        screen.getByRole("button", { name: /delete category/i })
      ).toBeInTheDocument();
    });

    it('a button "Close"', () => {
      expect(
        screen.getByRole("button", { name: /close/i })
      ).toBeInTheDocument();
    });

    it("error message when submitting a form with empty inputs", async () => {
      expect(screen.queryByText(/category is required/i)).toBeNull();

      userEvent.click(screen.getByRole("button", { name: /delete category/i }));

      await renderComponentAsync();

      expect(screen.getByText(/category is required/i)).toBeInTheDocument();
    });
  });

  it("successful form submission", async () => {
    await renderComponentAsync();

    userEvent.click(screen.getByRole("button", { name: /category ​/i }));
    userEvent.click(screen.getByRole("option", { name: /work/i }));

    userEvent.click(screen.getByRole("button", { name: /delete category/i }));

    await waitFor(() =>
      expect(onSubmit).toHaveBeenCalledWith({
        id: "084a3558-6b14-4e38-9d24-47e912021384",
      })
    );

    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
