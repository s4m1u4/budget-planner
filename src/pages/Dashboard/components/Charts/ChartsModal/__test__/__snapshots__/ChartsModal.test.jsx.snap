// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartsModal component snapshot matching 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-5fyqun"
      tabindex="-1"
    >
      <h3
        class="css-ngqp4j"
      >
        Delete category
      </h3>
      <form>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-14q6lbf-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            Category
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            variant="outlined"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="mui-1-label mui-1"
              class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              name="id"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Category
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="css-20ow0d"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-p8fncj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Delete category
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-kl9a8j-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
